### DISCUSSION TITLE
Is there a standard option of pass system environment option to SCons ENV?

### DISCUSSION THREAD

**Discussion by abitrolly** [SCons/scons#4270] [Q&A]
https://github.com/SCons/scons/pull/4239 proposes passing down reproducible builds var `SOURCE_DATE_EPOCH` automatically. I don't think it is a good way, but modifying `SConstruct` for passing it down explicitly, or making it optional, is also non-ideal. Is there a command line option for `scons` command that allows to pass systems environment variables to make them appear on SCons environment?

**Comment by bdbaddog** [comment]
This PR has been merged to enable such.. https://github.com/SCons/scons/pull/4261

There is no SCons defined command line option to do such, but anyone creating a build system with SCons can easily add one.

SCons has an explicit guarantee to not do so by default.

**Reply by abitrolly** [reply]
If I understand it right, #4261 contains modification to make SCons pass `SOURCE_DATE_EPOCH` to all builds on specific machine automatically. This was discussed in #4239, and it's not what I've meant.

Is there an example how to add the env passthrough option to the `SConstruct`?

**Comment by acmorrow** [comment]
@abitrolly - If you arrange your `SConstruct` so it can take a "variables file" argument, you can write a variables file which forces the shell environment into the SCons `ENV`:

- https://github.com/mongodb/mongo/blob/fc98977e569884357584111e19873a0da6bf7217/SConstruct#L608-L613
- https://github.com/mongodb/mongo/blob/fc98977e569884357584111e19873a0da6bf7217/SConstruct#L916-L936
- https://github.com/mongodb/mongo/blob/fc98977e569884357584111e19873a0da6bf7217/SConstruct#L1071-L1076
- https://github.com/mongodb/mongo/blob/fc98977e569884357584111e19873a0da6bf7217/SConstruct#L1551
- https://github.com/mongodb/mongo/blob/fc98977e569884357584111e19873a0da6bf7217/etc/scons/propagate_shell_environment.vars

Now `scons --variables-files=etc/scons/propagate_shell_environment.vars` gets you the shell environment in `ENV`.

**Reply by abitrolly** [reply]
I'd prefer more explicit syntax like `scons -e SOURCE_DATE_EPOCH=$SOURCE_DATE_EPOCH` (like in `docker run`).

**Reply by bdbaddog** [reply]
IMHO that'd voilate the no env promise. The build system creator would have propagated variables they want propagated and/or added their own switch.
I don't see this a change I'd be willing to merge.

**Reply by abitrolly** [reply]
So is there an example how build system creator can add this as an option?

**Reply by bdbaddog** [reply]
Have you read the docs? See the section on command line arguments in the users guide.
https://scons.org/doc/production/HTML/scons-user.html#sect-command-line-options

**Reply by abitrolly** [reply]
I mean the documentation about maing reproducible builds with SCons could include the example. When this thread will be indexed by search engines, example how to `SOURCE_DATE_EPOCH` down with the option will be useful.

**Reply by bdbaddog** [reply]
That's really beyond SCons' scope. See reproducible builds website for details on how to do so, then it's an implementation detail in any build system you use.

**Reply by abitrolly** [reply]
If dealing with `SOURCE_DATE_EPOCH` is beyond scope, then why #4261 got merged?

**Reply by mwichmann** [reply]
We can certainly add an example to the Cookbook (presented at https://scons-cookbook.readthedocs.io/en/latest/, repository at https://github.com/scons/cookbook).  We used to put lots of stuff on the wiki, but once the wiki moved to github, that doesn't help because search engines don't get to index that.

**Reply by bdbaddog** [reply]
Still out of scope.

There is a FAQ item related to passing shell env vars into Environment()..
https://scons.org/faq.html#Why_doesn.27t_SCons_find_my_compiler.2BAC8-linker.2BAC8-etc..3F_I_can_execute_it_just_fine_from_the_command_line.
